src/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -81,57 +81,141 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
     }
 }
 
-void CodeGen::genSSEIntrinsic(GenTreeHWIntrinsic* node)
+void CodeGen::genHWIntrinsic_R_R_RM(GenTreeHWIntrinsic* node, instruction ins)
 {
-    NamedIntrinsic intrinsicID = node->gtHWIntrinsicId;
-    GenTree*       op1         = node->gtGetOp1();
-    GenTree*       op2         = node->gtGetOp2();
-    regNumber      targetReg   = node->gtRegNum;
-    var_types      targetType  = node->TypeGet();
-    var_types      baseType    = node->gtSIMDBaseType;
+    var_types targetType = node->TypeGet();
+    regNumber targetReg  = node->gtRegNum;
+    GenTree*  op1        = node->gtGetOp1();
+    GenTree*  op2        = node->gtGetOp2();
+    emitter*  emit       = getEmitter();
+
+    // TODO-XArch-CQ: Commutative operations can have op1 be contained
+    // TODO-XArch-CQ: Non-VEX encoded instructions can have both ops contained
 
     regNumber op1Reg = op1->gtRegNum;
-    regNumber op2Reg = REG_NA;
-    emitter*  emit   = getEmitter();
+
+    assert(targetReg != REG_NA);
+    assert(op1Reg != REG_NA);
 
     genConsumeOperands(node);
 
+    if (op2->isContained() || op2->isUsedFromSpillTemp())
+    {
+        unsigned varNum = BAD_VAR_NUM;
+        unsigned offset = (unsigned)-1;
+
+        if (op2->isIndir())
+        {
+            assert(op2->isContained());
+
+            GenTreeIndir* mem     = op2->AsIndir();
+            GenTree*      memBase = mem->Base();
+
+            switch (memBase->OperGet())
+            {
+                case GT_LCL_VAR_ADDR:
+                {
+                    varNum = memBase->AsLclVarCommon()->GetLclNum();
```

Comment:
If `memBase` is a `LCL_VAR_ADDR`, it still may be the case that there are other components to the addressing mode (i.e. if the child of the `IND` is an `LEA`, then it could have all of the various addressing mode components, unless you've excluded those cases from containment, which I don't *think* you have.

Reply:
I had followed [emitInsBinary](`https://github.com/dotnet/coreclr/blob/master/src/jit/emitxarch.cpp#L2902`) as an example for the various emit paths that needed to be supported.

It looked like, the `GT_LCL_VAR_ADDR` just set the `varNum` and `offset` (https://github.com/dotnet/coreclr/blob/master/src/jit/emitxarch.cpp#L3059) and then it just calls `emitIns_R_S` (https://github.com/dotnet/coreclr/blob/master/src/jit/emitxarch.cpp#L3082).

I didn't think it was possible for `GetLclNum` to return `BAD_VAR_NUM` in this case, so the subsequent code path (which calls into `emitIns_R_M`) wasn't needed. Was that an incorrect assumption?

Reply:
Thanks. I think you're right, and it is probably the case that, if we have anything more than a base, we would use `GT_LCL_FLD_ADDR` instead. I think it would be prudent to add `assert(mem->Index() == nullptr && mem->Scale() == 1 && mem->Offset() == 0)`

Reply:
Ok, new commit should be validating that `varNum` and `offset` are correct for the general `emitIns_R_S` case and that `baseReg`, `indxReg`, `scale`, and `offs` are their default values for the `GT_LCL_VAR_ADDR` case.

Reply:
I have tests that will validate various scenarios that should end up with contained nodes here: https://github.com/dotnet/coreclr/pull/15771